recitation_2/makefile.md

```diff
@@ -0,0 +1,401 @@
+goodbye.h       hello.c         hello.o         myprogram.o
+```
+
+- `gcc -c` will pre-process (ie. including the headers), compile, and assemble the files; taking in a `.c` file and outputing a `.o` file.
+
+### Linking
+
+Now, we need to link the `.o` files to create our executable.
+
+To link:
+
+```bash
+$ gcc -o myprogram hello.o goodbye.o myprogram.o
+```
+
+Format: `gcc -o <output name> <ingredients>`
+
+Now we have our executable, `myprogram`:
+
+```bash
+$ ls
+goodbye.c       goodbye.o       hello.h         myprogram       myprogram.o
+goodbye.h       hello.c         hello.o         myprogram.c
+```
+
+```bash
+$ ./myprogram Faustina
+This is my program!
+Hello, Faustina! 
+Goodbye, Faustina!
+```
+
+### Makefile
+
+This process to produce our executable required many steps. We can streamline it by creating a `Makefile`, which gives us a simple way to organize code compilation.
+
+Because `hello.o` is produced by including `hello.h` and compiling `hello.c`, we can say that `hello.o` depends on `hello.c` and `hello.h`. Similarly, `myprogram` is produced by linking `hello.o`, `goodbye.o`, and `myprogram.o`, so we say that `myprogram` depends on these object files.
+
+We should only need to rebuild something if at least one of the files it depends on has changed. After all, part of the reason for separate compilation is to avoid the cost of rebuilding things we do not need to rebuild.
+
+Make is a build system that helps us manage incremental builds. You can find its manual online here: https://www.gnu.org/software/make/manual/make.html.
+
+Make runs commands for you according to instructions specified in a Makefile. For example, it can help you run any of the GCC build commands shown before. A `Makefile` should specify one or more “rules” for building files; each rule looks like this:
+
+```bash
+myprogram: hello.o goodbye.o myprogram.o
+	gcc -o myprogram hello.o goodbye.o myprogram.o
+```
+
+where:
+
+- `myprogram` is the name of the target file this rule will build
+- `hello.o`, `goodbye.o`, `myprogram.o` are the dependencies (prerequisites) that the target depends on
+- `gcc -o myprogram hello.o goodbye.o myprogram.o` is the recipe (ie. the shell command that builds the target)
+    - `goodbye.o` and `myprogram.o` are the ingredients for this recipe
+
+Make uses the “last modified” timestamp of each file to figure out if a target needs to be rebuilt. Make will rebuild a target if it doesn’t exist, or if it was last modified earlier than one of its dependencies.
+
+Creating our `Makefile` for `myprogram`:
+
+```bash
+myprogram: hello.o goodbye.o myprogram.o
+	gcc -o myprogram hello.o goodbye.o myprogram.o
+
+hello.o: hello.c hello.h
+	gcc -c -o hello.o hello.c
+
+goodbye.o: goodbye.c goodbye.h
+	gcc -c -o goodbye.o goodbye.c
+
+myprogram.o: myprogram.c hello.h goodbye.h
+	gcc -c -o myprogram.o myprogram.c
+```
+
+Note: The space preceding your recipe needs to be a tab, not spaces.
```

Comment:
```suggestion
Note: The space preceding your recipe needs to be a tab, not four spaces.
```

Reply:
They look the same, but 4 spaces won't work. It's necessary to emphasize makefile is very sensitive, so be mindful when configuring vimrc.